sweep/sweeper.go

```diff
@@ -29,13 +64,387 @@ type UtxoSweeperConfig struct {
+
+				// In case of a double spend, don't notify
+				// listeners yet but wait for the spending tx to
+				// trigger the spend notification.
+				if err == lnwallet.ErrDoubleSpend {
+					clearInputsToSweep()
+					break
+				}
+
+				// In case of an other error, signal all
+				// involved inputs the error.
+				for _, input := range inputToSweep {
+					s.signalResult(input.OutPoint(),
+						Result{Err: err})
+				}
+
+				// Do not try to sweep any of the inputs that
+				// cause this error again.
+				clearInputsToSweep()
+
+				break
+			}
+
+			// Store published transaction hash to detect remote
+			// spend when the spend event comes in.
+			sweepTxHash := sweepTx.TxHash()
+			for _, input := range inputToSweep {
+				outpoint := input.OutPoint()
+				s.pendingInputs[*outpoint].txHash = sweepTxHash
+			}
+
+			// Clear all pending inputs as these have now been
+			// published successfully.
+			clearInputsToSweep()
+		case <-s.quit:
+			return
+		}
+
+		// TODO: Handle the case where a tx is not confirmed within
+		// specified time and fee needs to be bumped.
+	}
+
+}
+
+func (s *UtxoSweeper) signalResult(outpoint *wire.OutPoint, result Result) {
+	pendInput := s.pendingInputs[*outpoint]
+	listeners := pendInput.listeners
+
+	if result.Err == nil {
+		log.Debugf("Dispatching sweep success for %v to %v listeners",
+			outpoint, len(listeners))
+	} else {
+		log.Debugf("Dispatching sweep error for %v to %v listeners: %v",
+			outpoint, len(listeners), result.Err)
+	}
+
+	for _, resultChan := range listeners {
+		s.signalResultChan(resultChan, result)
+	}
+
+	// Cancel spend notification with chain notifier. This is not necessary
+	// in case of a success, except for that a reorg could still happen.
+	pendInput.ntfnRegCancel()
+
+	// Inputs are no longer pending after result has been sent.
+	delete(s.pendingInputs, *outpoint)
+}
+
+func (s *UtxoSweeper) signalResultChan(resultChan chan Result, result Result) {
+	select {
+	case resultChan <- result:
+	case <-s.quit:
+	case <-time.After(sweeperTimeout):
+		log.Errorf("Cannot deliver sweep result %v", result.Err)
+	}
+}
+
+func (s *UtxoSweeper) sweep(inputs []Input) (*wire.MsgTx, error) {
+
+	// If nothing to sweep, then continue waiting for next
+	// block.
+	if len(inputs) == 0 {
+		return nil, nil
+	}
+
+	_, currentHeight, err := s.cfg.ChainIO.GetBestBlock()
+	if err != nil {
+		return nil, err
+	}
+
+	tx, err := s.CreateSweepTx(inputs, s.cfg.SweepTxConfTarget,
+		uint32(currentHeight))
+	if err != nil {
+		return nil, fmt.Errorf("cannot create sweep tx: %v", err)
 	}
+
+	// Publish sweep tx.
+
+	log.Debugf("Publishing sweep tx")
+	err = s.cfg.PublishTransaction(tx)
+	if err != nil {
+		return nil, err
+	}
+
+	return tx, nil
+}
+
+func (s *UtxoSweeper) waitForSpend(input Input) (func(), error) {
+	outpoint := input.OutPoint()
+
+	log.Debugf("Wait for spend for %v", outpoint)
+
+	spendEvent, err := s.cfg.Notifier.RegisterSpendNtfn(
+		outpoint,
+		input.SignDesc().Output.PkScript, 0)
```

Comment:
Style nit: can pack these all on a single line looks like, then with a trailing comma and close paren on a new line.